Assignment3/PointAndGrid.py

```diff
@@ -0,0 +1,280 @@
+        """
+        Returns number of columns. Size of a row.
+        :return: int
+        """
+        return len(self._grid[0]) if self._grid is not None else 0
+
+    def verticalSize(self):
+        """
+        Returns number of rows. Size of a column.
+        :return: int
+        """
+        return len(self._grid) if self._grid is not None else 0
+
+    def getElement(self, tuple):
+        """
+        Return an element with the given coordinates.
+        :param tuple:
+        :return: any type
+        """
+        row = tuple[0]
+        col = tuple[1]
+        return self._grid[row][col]
+
+    def findElementInGrid(self, element):
+        """
+        Return the list of all Points with the grid and coordinates of the element if exists. None otherwise.
+        :param element: any type
+        :return: Point
+        """
+        result = []
+        R, C = self.getSize()
+        for r in range(R):
+            for c in range(C):
+                if self._grid[r][c] == element:
+                    result.append(Point(self, (r, c), element))
+        return result
+
+    def findNeighborsOfPoint(self, tuple):
+        """
+        Return list of Points that are neighbours of element with the given coordinates.
+        :param tuple: Tuple
+        :return: list(Point) or None
+        """
+        r = tuple[0]
+        c = tuple[1]
+        result = []
+        possible_coord = [(r-1, c-1),(r-1, c),(r, c-1),(r-1, c+1),
+                          (r+1, c-1),(r+1, c),(r, c+1),(r+1, c+1)]
+
+        for coord in possible_coord:
+            if self._validCoordinates(coord):
+                result.append(Point(self, coord, self.getElement(coord)))
+        return result if result else None
+
+    def ifWordInGrid(self, word):
+        """
+        Return True if the word can be formed in grid, False otherwise.
+        :param word: str
+        :return: bool
+        """
+        word = word.lower()
+        return self.findString(word) is not None
+
+    ############################## THE SOLUTION TO ASSIGNMENT 3 ##########################################
+
+    def findString(self, word):
+        """
+        Find the word in the grid where we can starting from one position move to any neighbors,
+        but can't visit the same cell twice. Return the path if exists, None otherwise.
+        :param word: string
+        :return: list(Point) or None
+        """
+        startPoints = self.findElementInGrid(word[0])
+        path = []
+        for startP in startPoints:
+            # we are here only if the first letter of word is in grid
+            path = [startP]
+            path = self._findPathRecursively(startP, word, path, [])
+        if self._pathIsWord(path, word):
+            return path
+
+    def _pathIsWord(self, path, word):
+        if not path:
+            return False
+        return [p.getValue() for p in path] == list(word)
+
+    def _findPathRecursively(self, point, word, path, visited):

```

Comment:
So does this work now? It still looks a bit suspicious though I'm not sure by just reading it. I would expect something like this to produce recursively find words:

```
def _findPathRecursively(self, point, word, path):
  # Recurse into the point.
  path = list(path)
  path.append(point)
  if self._pathIsWord(path, word):
    # Success! Propagate the result up!
    return path

  possible_ways = self.findPossibleWays(path, point, word)
  for way in possible_ways:
    result = self._findPathRecursively(way, word, path)
    if result:
      # Somewhere down the recursion tree a correct result was found! Propagate it up!
      return result
  # None of the recursions yield the word we look for, signal failure up the stack.
  return None
````

And when starting the recursion:

```
path = self._findPathRecursively(startP, word, [])
if path:
  return path
```

I'm sure the above is not exactly correct but I hope it illustrates how recursive search should return successful and failed recursion up the stack. The lack of these return signals in this code is what makes me suspect it shouldn't yield correct results.